worktree_status.go

```diff
@@ -243,30 +251,170 @@ func diffTreeIsEquals(a, b noder.Hasher) bool {
 }
 
 // Add adds the file contents of a file in the worktree to the index. if the
-// file is already staged in the index no error is returned.
+// file is already staged in the index no error is returned. If a file deleted
+// from the Workspace is given, the file is removed from the index.
 func (w *Worktree) Add(path string) (plumbing.Hash, error) {
 	s, err := w.Status()
 	if err != nil {
 		return plumbing.ZeroHash, err
 	}
 
-	h, err := w.copyFileToStorage(path)
+	idx, err := w.r.Storer.Index()
+	if err != nil {
+		return plumbing.ZeroHash, err
+	}
+
+	added, h, err := w.doAdd(idx, s, path)
+	if err != nil {
+		return h, err
+	}
+
+	if !added {
+		return h, nil
+	}
+
+	return h, w.r.Storer.SetIndex(idx)
+}
+
+// AddDirectory adds the files contents of a directory and all his
+// sub-directories recursively in the worktree to the index. If any of the
+// file is already staged in the index no error is returned.
+func (w *Worktree) AddDirectory(path string) error {
```

Comment:
This API seems a little clumsy to me. Is there a reason we can simply detect and support directory paths as part of the current `Add` function? A `path string` is already sufficient for a user to describe either directories or files.

Moreover, if we are going to provide separate functions for directories on `Add` we should also include separate functions `MoveDirectory` and `RemoveDirectory` for consistency.

In any case, I think we should simply extend the existing `Add`, `Move`, and `Remove` to support directory paths.

Reply:
what should I return when I `Add` a repository? Because the signature returns a `Hash`.

Reply:
It seems like `Add` returns the hash of the blob object that was added to the repository. If one is adding a directory, it might make sense to return the hash of the corresponding tree object.

Reply:
`Add` doesn't create any tree object.

Reply:
Ah yes. I see that. I guess this works then.